versioned/spi/src/main/java/org/projectnessie/versioned/TracingVersionStore.java

```diff
@@ -0,0 +1,334 @@
+    try {
+      result = delegate.handle().onClose(scope::close);
+      return result;
+    } catch (IllegalArgumentException e) {
+      // IllegalArgumentException is a special kind of exception that indicates a user-error.
+      throw e;
+    } catch (RuntimeException e) {
+      throw traceRuntimeException(scope, e);
+    } finally {
+      // See below (delegateStream1Ex)
+      if (result == null) {
+        scope.close();
+      }
+    }
+  }
+
+  private <R, E1 extends VersionStoreException> Stream<R> delegateStream1Ex(String spanName, Consumer<SpanBuilder> spanBuilder,
+      DelegateWith1<Stream<R>, E1> delegate) throws E1 {
+    Scope scope = createActiveScope(spanName, spanBuilder);
+    Stream<R> result = null;
+    try {
+      result = delegate.handle().onClose(scope::close);
+      return result;
+    } catch (IllegalArgumentException e) {
+      // IllegalArgumentException is a special kind of exception that indicates a user-error.
+      throw e;
+    } catch (RuntimeException e) {
+      throw traceRuntimeException(scope, e);
+    } finally {
+      // We cannot `catch (E1 e)`, so assume that the delegate threw an exception, when result==null
+      // and then close the trace-scope.
+      if (result == null) {
+        scope.close();
+      }
+    }
+  }
+
+  private <R> R delegate(String spanName, Consumer<SpanBuilder> spanBuilder, Delegate<R> delegate) {
+    try (Scope scope = createActiveScope(spanName, spanBuilder)) {
+      try {
+        return delegate.handle();
+      } catch (IllegalArgumentException e) {
+        // IllegalArgumentException is a special kind of exception that indicates a user-error.
+        throw e;
+      } catch (RuntimeException e) {
+        throw traceRuntimeException(scope, e);
+      }
+    }
+  }
+
+  private <R, E1 extends VersionStoreException> R delegate1Ex(String spanName, Consumer<SpanBuilder> spanBuilder,
+      DelegateWith1<R, E1> delegate) throws E1 {
+    try (Scope scope = createActiveScope(spanName, spanBuilder)) {
+      try {
+        return delegate.handle();
+      } catch (IllegalArgumentException e) {
+        // IllegalArgumentException is a special kind of exception that indicates a user-error.
+        throw e;
+      } catch (RuntimeException e) {
+        throw traceRuntimeException(scope, e);
+      }
+    }
+  }
+
+  private <E1 extends VersionStoreException, E2 extends VersionStoreException> void delegate2Ex(String spanName,
+      Consumer<SpanBuilder> spanBuilder, DelegateWith2<E1, E2> delegate) throws E1, E2 {
+    try (Scope scope = createActiveScope(spanName, spanBuilder)) {
+      try {
+        delegate.handle();
+      } catch (IllegalArgumentException e) {
+        // IllegalArgumentException is a special kind of exception that indicates a user-error.
+        throw e;
+      } catch (RuntimeException e) {
+        throw traceRuntimeException(scope, e);
+      }
+    }
+  }
+
+  @FunctionalInterface
+  interface Delegate<R> {
```

Comment:
The overloading of `Delegate`, `delegate()` and `delegate` is slightly confusing. Could we rename a few to create more distinction?

Reply:
Renamed some of those.